EZP-27271: Unable to upload large binary file using REST / Platform UI

kmadejski opened the pull request:
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-27271](https://jira.ez.no/browse/EZP-27271)
| **Bug/Improvement**| no
| **New feature**    | no
| **Target version** | `6.7`
| **BC breaks**      | no
| **Tests pass**     | yes/no
| **Doc needed**     | no

This PR contains small refactoring in `eZ\Publish\Core\REST\Server\Controller\Content` which allows PlatformUI to reuse once parsed Request as `RestContentCreateStruct`.

Related PlatformUI PR: https://github.com/ezsystems/PlatformUIBundle/pull/953

**TODO**:
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

adamwojs commented on eZ/Publish/Core/REST/Server/Controller/Content.php:
```diff
@@ -736,4 +705,51 @@ protected function forward($controller)
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     * @param \eZ\Publish\Core\REST\Server\Values\RestContentCreateStruct $contentCreate
+     *
+     * @throws NotFoundException
```
Nitpick: Please use FQCN

andrerom commented:
@kmadejski Seems you are addding stuff here after review or?
As in where is the change on `convert()` in ae57ff8 used? Seems it is not really streaming the research but buffering it all into memory so maybe that concept need some further work in separate PR or? _(afaik only Xml via XMLReader would be able to act on stream afaik, so you don't need to buffer the string and the xml document in memory at the same time)_

kmadejski commented:
@andrerom yes, I've changed small nitpick and amended commit after that. I've updated PR description with correct commit URL. 
Secondly, yes, I've added a change in `convert()` method later, when I realized that it might be a good way to decrease memory usage. 

> As in where is the change on convert() in ae57ff8 used?

The change was introduced to be able to benefit from `$request->getContent(true)`: https://github.com/ezsystems/PlatformUIBundle/pull/953/files#diff-509a29faee83ecf7375c7176a93f90afR40
Thanks to that we don't need to store duplicated request body in the memory.

andrerom commented:
Issues:
-  Message->body is still documented as string, might want to reconsider design here, just documenting it taking both types all over is code smell for not solving the design in a good way, don't have a good suggestion right now but we can try to figure that out.
- Code here does not use the same trick, as in `$request->getContent()` is still using string here
- `is_resource($input) ? $dom->loadXML(stream_get_contents($input)) : $dom->loadXML($input);` still implies content will exist twice _(down from 3+)_, first both as string & DOM, and afterwards as DOM & parsed array until parsing is done. While not something we can do anything about for json, for XML we can potentially do it as a stream: https://stackoverflow.com/a/18520386/505567


@kmadejski What if you 1. move `$contentCreate = $this->inputDispatcher->parse(` into own `function parseCreateContent(Request $request)`  and reuse that from PlatformUI so that is independent of type choice, and 2. then move resource changes to new PR so we can merge the fix for double parsing first and deal with resource usage alone as step two?

kmadejski commented:
@andrerom PR changed as requested 🙂 

ping @adamwojs, @andrerom, @alongosz do the review once again, please.

kmadejski commented:
@andrerom yes, that approach makes sense too. We will see what @adamwojs & @alongosz think about this.

alongosz commented:
@kmadejski please see [CI failues](https://travis-ci.org/ezsystems/ezpublish-kernel/jobs/353290341#L1503) :trollface:

kmadejski commented:
@alongosz as I said before, I believe that these tests fail because changes from https://github.com/ezsystems/PlatformUIBundle/pull/953 need to be merged first.

The error says: `Compile Error: Access level to EzSystems\PlatformUIBundle\Controller\Rest\ContentController::parseRequestContent() must be protected (as in class eZ\Publish\Core\REST\Server\Controller\Content) or weaker` but that method was moved from `EzSystems\PlatformUIBundle\Controller\Rest\ContentController` to `eZ\Publish\Core\REST\Server\Controller\Content`.

andrerom commented:
@kmadejski could you pick a different (new) name for `parseRequestContent`? I don't want people updating and for some reason only getting kernel update hitting this. For instance `parseCreateContentRequest` would be more correct anyway right?

kmadejski commented:
@andrerom In the beginning, I thought that this method will be used also in other methods like `updateContentMetadata`, `udpateVersion` or `createRelation` but it seems that these methods use different headers (passed to the `Message` constructor). So, I've changed the method name as you proposed and it looks good now.

alongosz commented:
> For instance parseCreateContentRequest would be more correct anyway right?

Seems like a better approach :)
But that better naming gave me a random thought - @kmadejski what if someone uploads large file on Content update? PR covers Content create only.

kmadejski commented:
@alongosz thankfully, `updateVersion` wasn't affected. There is no duplicated request content parsing. Request content is parsed and dispatched only once in the kernel: `eZ\Publish\Core\REST\Server\Controller\Content::updateVersion()`. Take a glance here:
https://github.com/ezsystems/PlatformUIBundle/blob/master/Controller/Rest/ContentController.php#L89-#L101

alongosz commented:
Ok, that's great :)

andrerom commented on eZ/Publish/Core/REST/Server/Controller/Content.php:
```diff
@@ -736,4 +702,66 @@ protected function forward($controller)
+
+        $contentValue = null;
+        $contentType = null;
+        $relations = null;
```
This is some light Friday OCD kicking in, but this could also be written as:
` $contentValue = $contentType = $relations = null;`

or 
```
$relations = null;
$contentType = null;
$contentValue = null;
```

;)
